### PR TITLE
Turn+concede

### PR DESCRIPTION
**Pull request by dtrypt**
Game is in a playable state now!

- Added the turn-based logic.
- When it's a player's turn they can no longer move pieces of the opposite color.
- Players can't move pieces when its not their turn.
- Added concede button allowing players to forfeit and end the game.
- Pieces cannot be moved anymore in finished games.
- Categorized the games on the homepage into: my games - finished games - available games - full games.
(homepage now acts kind of like a user dashboard without a scoring system)
- Not sure why the tests in the chess controller are failing, I manually tested them and everything seemed to work correctly. Didn't run Rubocop.
- Added a logo and flavicon.
- Facebook login should now work on Heroku.

- Reset your local database and run the migration to account for database changes.
- Try to find bugs if you can!

*Update: I didn't use rubocop here because I'm having some issues with it locally. It's picking up errors in files that haven't been touched in a long time and when I go to the file to fix the error, I don't see it. I spent a good amount of time trying to debug rubocop but haven't prevailed. There's also complexity issues and I'm a little strapped for time so I didn't bother. If anyone wants to help out with this, you can push commits to this branch. Otherwise, I'll come back and finish it off later on, but for the meantime, if there's no bugs in the code that affect functionality, I'd like to push it to production.

### CODE REVIEW

**Comment by sungwon** [comment]
lol @ "Didn't run Rubocop"

**Comment by dtrypt** [comment]
@sungwon lol..yea I just don't have time right now. I updated the PR description to address not using rubo.

**Comment by sungwon** [comment]
@trustinruss I just thought it was funny since Rubocop can be pretty annoying. 

I haven't had time to look at the Turn-based commit in depth, but the other commits look good.

**Comment by dtrypt** [comment]
@sungwon Yea I know =P, just figured I should add a little explanation. Thanks for still checking our pull requests!

**Review by LesleyPark — COMMENTED** [review]
Great job on this Russ! A couple things:

-Looks like adding the turn based logic also brings back that issue with captures where you have two pieces on one square while the game refreshes
-I was able to create a checkmate situation but it looks like that doesn't cause the game to move into the finished games section you created. Forfeited games seem to though.

Aside from this, everything seems to be working like it should.

**Comment by dtrypt** [comment]
@LesleyPark Nice catches! The issues should be fixed now.

**Comment by LesleyPark** [comment]
Looks good. Any idea what's going on with the failing tests? I'm having similar issues with stalemate as well.

File: app/controllers/chess_controller.rb [hunk 1]
```diff
@@ -1,7 +1,17 @@
 class ChessController < ApplicationController
   def index
+    @games = Game.all
+    @my_games = []
+    @finished_games = []
+    if current_user
+      @games.each do |game|
+        @my_games << game if (game.white_player == current_user.id || game.black_player == current_user.id) && game.finished == false
```
Comment (koglesby) [inline app/controllers/chess_controller.rb]: You can reduce the cyclomatic complexity by moving some of these conditions into a new method. Also, I think it makes more sense to have finished games only show up in the finished games section, and not the full games section.

        @my_games << game if game_player(game) && !game.finished
        @finished_games << game if game_player(game) && game.finished
      end
    end
    @available_games = Game.available
    @full_games = @games - @available_games - @finished_games
    flash[:notice] = nil
    flash[:alert] = nil
  end

  private

  def game_player(game)
    (game.white_player == current_user.id || game.black_player == current_user.id)
  end

File: app/controllers/games_controller.rb [hunk 2]
```diff
@@ -26,8 +26,14 @@ def show
 
   def update
     @game = Game.find(params[:id])
-    @game.update_attributes(black_player: current_user.id) if @game.black_player.nil?
-    redirect_to game_path(@game)
+    if @game.black_player.nil? && params[:concede].blank?
+      @game.update_attributes(black_player: current_user.id)
+      return redirect_to game_path(@game)
+    end
+    if params[:concede] == 'true'
```
Comment (koglesby) [inline app/controllers/games_controller.rb]: return if params[:concede]
    @game.update_attributes(finished: true)
    redirect_to root_path

Fixes rubocop "gaurd clause" offense

File: app/controllers/games_controller.rb [hunk 3]
```diff
@@ -38,9 +44,20 @@ def game_params
 
   def flash_notices
     flash[:notice] = nil
-    return flash[:notice] = 'Checkmate. Black Player Wins!' if @game.checkmate!('white')
-    return flash[:notice] = 'Checkmate. White Player Wins!' if @game.checkmate!('black')
-    flash[:notice] = 'Black King Check!' if @game.check?('black')
-    flash[:notice] = 'White King Check!' if @game.check?('white')
+    flash[:alert] = nil
+    flash[:notice] = 'White Player\'s turn' if @game.turn
+    flash[:notice] = 'Black Player\'s turn' if @game.turn == false
```
Comment (koglesby) [inline app/controllers/games_controller.rb]: flash[:notice] = @game.turn ? 'White Player\'s turn' : 'Black Player\'s turn'

Reduces cyclomatic complexity

File: app/controllers/pieces_controller.rb [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
-    respond_to do |format|
-      format.html { redirect_to game_path(@piece.game) }
-      format.json { render json: @piece }
+    if validate_player
+      success = @piece.move_to!(piece_params)
+      @piece.game.update_attributes(turn: change_turn(@piece.color)) if success
+      respond_to do |format|
+        format.html { redirect_to game_path(@piece.game) }
+        format.json { render json: @piece }
+      end
```
Comment (koglesby) [inline app/controllers/pieces_controller.rb]: return unless validate_player
    success = @piece.move_to!(piece_params)
    @piece.game.update_attributes(turn: change_turn(@piece.color)) if success
    respond_to do |format|
      format.html { redirect_to game_path(@piece.game) }
      format.json { render json: @piece }
    end

fixes rubocop "use gaurd clause" offense

File: app/helpers/games_helper.rb [hunk 5]
```diff
@@ -9,7 +9,9 @@ def locate_piece(x, y)
...
   end
 
-  # def path_to_king
-  #   return []
-  # end
+  def colors_turn(color)
+    return 'drag' if @game.turn == true && color == 'white' && !@game.finished
+    return 'drag' if !@game.turn && color == 'black' && !@game.finished
+    'default'
+  end
```
Comment (koglesby) [inline app/helpers/games_helper.rb]: if !@game.finished
      return 'drag' if (@game.turn && color == 'white') || (!@game.turn && color == 'black')
    else
      'default'
    end

reduces cyclomatic complexity

File: config/initializers/devise.rb [hunk 6]
```diff
@@ -248,8 +248,7 @@
   # Add a new OmniAuth provider. Check the wiki for more information on setting
   # up on your models and hooks.
   # config.omniauth :github, 'APP_ID', 'APP_SECRET', scope: 'user,public_repo'
-  config.omniauth :facebook, '1269930473093839', '7e159c833a2a86fde2c6d4d4586b3a9b', display: 'popup',
-                                                                                     callback_url: 'http://localhost:3030/users/auth/facebook/callback'
+  config.omniauth :facebook, ENV["FACEBOOK_KEY"], ENV["FACEBOOK_SECRET"], display: 'popup'
```
Comment (koglesby) [inline config/initializers/devise.rb]: simple change to single quotes fixes rubocop offense

**Review by koglesby — COMMENTED** [review]
Great job on this PR! I just made some suggestions on how to clean it up a bit and fix those rubocop offenses.

**Review by LesleyPark — APPROVED** [review]

**Pull request merged**